File: binding/src/printsupport/libqprinter.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qprinter.html
pub const qprinter = struct {

    /// New constructs a new QPrinter object.
    pub fn New() QtC.QPrinter {
        return qtc.QPrinter_new();
    }


    /// New2 constructs a new QPrinter object.
    pub fn New2(printer: ?*anyopaque) QtC.QPrinter {
        return qtc.QPrinter_new2(@ptrCast(printer));
    }


    /// New3 constructs a new QPrinter object.
    pub fn New3(mode: i32) QtC.QPrinter {
        return qtc.QPrinter_new3(@intCast(mode));
    }


    /// New4 constructs a new QPrinter object.
    pub fn New4(printer: ?*anyopaque, mode: i32) QtC.QPrinter {
        return qtc.QPrinter_new4(@ptrCast(printer), @intCast(mode));
    }


    pub fn DevType(self: ?*anyopaque, ) i32 {
        return qtc.QPrinter_DevType(@ptrCast(self));
    }

    /// Allows for overriding the related default method
    pub fn OnDevType(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) i32) void {
        qtc.QPrinter_OnDevType(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseDevType(self: ?*anyopaque, ) i32 {
        return qtc.QPrinter_QBaseDevType(@ptrCast(self));
    }

    pub fn SetOutputFormat(self: ?*anyopaque, format: i32) void {
        qtc.QPrinter_SetOutputFormat(@ptrCast(self), @intCast(format));
    }

    pub fn OutputFormat(self: ?*anyopaque, ) i32 {
        return qtc.QPrinter_OutputFormat(@ptrCast(self));
    }

    pub fn SetPdfVersion(self: ?*anyopaque, version: i32) void {
        qtc.QPrinter_SetPdfVersion(@ptrCast(self), @intCast(version));
    }

    pub fn PdfVersion(self: ?*anyopaque, ) i32 {
        return qtc.QPrinter_PdfVersion(@ptrCast(self));
    }

    pub fn SetPrinterName(self: ?*anyopaque, printerName: []const u8) void {
        const printerName_str = qtc.libqt_string{
    .len = printerName.len,
    .data = printerName.ptr,
};
qtc.QPrinter_SetPrinterName(@ptrCast(self), printerName_str);
    }

    pub fn PrinterName(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QPrinter_PrinterName(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qprinter.PrinterName: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn IsValid(self: ?*anyopaque, ) bool {
        return qtc.QPrinter_IsValid(@ptrCast(self));
    }

    pub fn SetOutputFileName(self: ?*anyopaque, outputFileName: []const u8) void {
        const outputFileName_str = qtc.libqt_string{
    .len = outputFileName.len,
    .data = outputFileName.ptr,
};
qtc.QPrinter_SetOutputFileName(@ptrCast(self), outputFileName_str);
    }

    pub fn OutputFileName(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QPrinter_OutputFileName(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qprinter.OutputFileName: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetPrintProgram(self: ?*anyopaque, printProgram: []const u8) void {
        const printProgram_str = qtc.libqt_string{
    .len = printProgram.len,
    .data = printProgram.ptr,
};
qtc.QPrinter_SetPrintProgram(@ptrCast(self), printProgram_str);
    }

    pub fn PrintProgram(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QPrinter_PrintProgram(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qprinter.PrintProgram: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetDocName(self: ?*anyopaque, docName: []const u8) void {
        const docName_str = qtc.libqt_string{
    .len = docName.len,
    .data = docName.ptr,
};
qtc.QPrinter_SetDocName(@ptrCast(self), docName_str);
    }

    pub fn DocName(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QPrinter_DocName(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qprinter.DocName: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetCreator(self: ?*anyopaque, creator: []const u8) void {
        const creator_str = qtc.libqt_string{
    .len = creator.len,
    .data = creator.ptr,
};
qtc.QPrinter_SetCreator(@ptrCast(self), creator_str);
    }

    pub fn Creator(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QPrinter_Creator(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qprinter.Creator: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetPageOrder(self: ?*anyopaque, pageOrder: i32) void {
        qtc.QPrinter_SetPageOrder(@ptrCast(self), @intCast(pageOrder));
    }

    pub fn PageOrder(self: ?*anyopaque, ) i32 {
        return qtc.QPrinter_PageOrder(@ptrCast(self));
    }

    pub fn SetResolution(self: ?*anyopaque, resolution: i32) void {
        qtc.QPrinter_SetResolution(@ptrCast(self), resolution);
    }

    pub fn Resolution(self: ?*anyopaque, ) i32 {
        return qtc.QPrinter_Resolution(@ptrCast(self));
    }

    pub fn SetColorMode(self: ?*anyopaque, colorMode: i32) void {
        qtc.QPrinter_SetColorMode(@ptrCast(self), @intCast(colorMode));
    }

    pub fn ColorMode(self: ?*anyopaque, ) i32 {
        return qtc.QPrinter_ColorMode(@ptrCast(self));
    }

    pub fn SetCollateCopies(self: ?*anyopaque, collate: bool) void {
        qtc.QPrinter_SetCollateCopies(@ptrCast(self), collate);
    }

    pub fn CollateCopies(self: ?*anyopaque, ) bool {
        return qtc.QPrinter_CollateCopies(@ptrCast(self));
    }

    pub fn SetFullPage(self: ?*anyopaque, fullPage: bool) void {
        qtc.QPrinter_SetFullPage(@ptrCast(self), fullPage);
    }

    pub fn FullPage(self: ?*anyopaque, ) bool {
        return qtc.QPrinter_FullPage(@ptrCast(self));
    }

    pub fn SetCopyCount(self: ?*anyopaque, copyCount: i32) void {
        qtc.QPrinter_SetCopyCount(@ptrCast(self), copyCount);
    }

    pub fn CopyCount(self: ?*anyopaque, ) i32 {
        return qtc.QPrinter_CopyCount(@ptrCast(self));
    }

    pub fn SupportsMultipleCopies(self: ?*anyopaque, ) bool {
        return qtc.QPrinter_SupportsMultipleCopies(@ptrCast(self));
    }

    pub fn SetPaperSource(self: ?*anyopaque, paperSource: i32) void {
        qtc.QPrinter_SetPaperSource(@ptrCast(self), @intCast(paperSource));
    }

    pub fn PaperSource(self: ?*anyopaque, ) i32 {
        return qtc.QPrinter_PaperSource(@ptrCast(self));
    }

    pub fn SetDuplex(self: ?*anyopaque, duplex: i32) void {
        qtc.QPrinter_SetDuplex(@ptrCast(self), @intCast(duplex));
    }

    pub fn Duplex(self: ?*anyopaque, ) i32 {
        return qtc.QPrinter_Duplex(@ptrCast(self));
    }

    pub fn SupportedResolutions(self: ?*anyopaque, ) []const u8 {
        return qtc.QPrinter_SupportedResolutions(@ptrCast(self));
    }

    pub fn SupportedPaperSources(self: ?*anyopaque, ) []const u8 {
        return qtc.QPrinter_SupportedPaperSources(@ptrCast(self));
    }

    pub fn SetFontEmbeddingEnabled(self: ?*anyopaque, enable: bool) void {
        qtc.QPrinter_SetFontEmbeddingEnabled(@ptrCast(self), enable);
    }

    pub fn FontEmbeddingEnabled(self: ?*anyopaque, ) bool {
        return qtc.QPrinter_FontEmbeddingEnabled(@ptrCast(self));
    }

    pub fn PaperRect(self: ?*anyopaque, param1: i32) QtC.QRectF {
        return qtc.QPrinter_PaperRect(@ptrCast(self), @intCast(param1));
    }

    pub fn PageRect(self: ?*anyopaque, param1: i32) QtC.QRectF {
        return qtc.QPrinter_PageRect(@ptrCast(self), @intCast(param1));
    }

    pub fn PrinterSelectionOption(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QPrinter_PrinterSelectionOption(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qprinter.PrinterSelectionOption: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetPrinterSelectionOption(self: ?*anyopaque, printerSelectionOption: []const u8) void {
        const printerSelectionOption_str = qtc.libqt_string{
    .len = printerSelectionOption.len,
    .data = printerSelectionOption.ptr,
};
qtc.QPrinter_SetPrinterSelectionOption(@ptrCast(self), printerSelectionOption_str);
    }

    pub fn NewPage(self: ?*anyopaque, ) bool {
        return qtc.QPrinter_NewPage(@ptrCast(self));
    }

    /// Allows for overriding the related default method
    pub fn OnNewPage(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) bool) void {
        qtc.QPrinter_OnNewPage(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseNewPage(self: ?*anyopaque, ) bool {
        return qtc.QPrinter_QBaseNewPage(@ptrCast(self));
    }

    pub fn Abort(self: ?*anyopaque, ) bool {
        return qtc.QPrinter_Abort(@ptrCast(self));
    }

    pub fn PrinterState(self: ?*anyopaque, ) i32 {
        return qtc.QPrinter_PrinterState(@ptrCast(self));
    }

    pub fn PaintEngine(self: ?*anyopaque, ) QtC.QPaintEngine {
        return qtc.QPrinter_PaintEngine(@ptrCast(self));
    }

    /// Allows for overriding the related default method
    pub fn OnPaintEngine(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) QtC.QPaintEngine) void {
        qtc.QPrinter_OnPaintEngine(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBasePaintEngine(self: ?*anyopaque, ) QtC.QPaintEngine {
        return qtc.QPrinter_QBasePaintEngine(@ptrCast(self));
    }

    pub fn PrintEngine(self: ?*anyopaque, ) QtC.QPrintEngine {
        return qtc.QPrinter_PrintEngine(@ptrCast(self));
    }

    pub fn SetFromTo(self: ?*anyopaque, fromPage: i32, toPage: i32) void {
        qtc.QPrinter_SetFromTo(@ptrCast(self), fromPage, toPage);
    }

    pub fn FromPage(self: ?*anyopaque, ) i32 {
        return qtc.QPrinter_FromPage(@ptrCast(self));
    }

    pub fn ToPage(self: ?*anyopaque, ) i32 {
        return qtc.QPrinter_ToPage(@ptrCast(self));
    }

    pub fn SetPrintRange(self: ?*anyopaque, range: i32) void {
        qtc.QPrinter_SetPrintRange(@ptrCast(self), @intCast(range));
    }

    pub fn PrintRange(self: ?*anyopaque, ) i32 {
        return qtc.QPrinter_PrintRange(@ptrCast(self));
    }

    pub fn Metric(self: ?*anyopaque, param1: i32) i32 {
        return qtc.QPrinter_Metric(@ptrCast(self), @intCast(param1));
    }

    /// Allows for overriding the related default method
    pub fn OnMetric(self: ?*anyopaque, callback: *const fn (?*anyopaque, i32) callconv(.c) i32) void {
        qtc.QPrinter_OnMetric(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseMetric(self: ?*anyopaque, param1: i32) i32 {
        return qtc.QPrinter_QBaseMetric(@ptrCast(self), @intCast(param1));
    }

    pub fn SetEngines(self: ?*anyopaque, printEngine: ?*anyopaque, paintEngine: ?*anyopaque) void {
        qtc.QPrinter_SetEngines(@ptrCast(self), @ptrCast(printEngine), @ptrCast(paintEngine));
    }

    /// Allows for overriding the related default method
    pub fn OnSetEngines(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        qtc.QPrinter_OnSetEngines(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseSetEngines(self: ?*anyopaque, printEngine: ?*anyopaque, paintEngine: ?*anyopaque) void {
        qtc.QPrinter_QBaseSetEngines(@ptrCast(self), @ptrCast(printEngine), @ptrCast(paintEngine));
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QPrinter_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/printer.html#types
pub const enums = struct {
    pub const PrinterMode = enum {
        pub const ScreenResolution: i32 = 0;
        pub const PrinterResolution: i32 = 1;
        pub const HighResolution: i32 = 2;
    };

    pub const PageOrder = enum {
        pub const FirstPageFirst: i32 = 0;
        pub const LastPageFirst: i32 = 1;
    };

    pub const ColorMode = enum {
        pub const GrayScale: i32 = 0;
        pub const Color: i32 = 1;
    };

    pub const PaperSource = enum {
        pub const OnlyOne: i32 = 0;
        pub const Lower: i32 = 1;
        pub const Middle: i32 = 2;
        pub const Manual: i32 = 3;
        pub const Envelope: i32 = 4;
        pub const EnvelopeManual: i32 = 5;
        pub const Auto: i32 = 6;
        pub const Tractor: i32 = 7;
        pub const SmallFormat: i32 = 8;
        pub const LargeFormat: i32 = 9;
        pub const LargeCapacity: i32 = 10;
        pub const Cassette: i32 = 11;
        pub const FormSource: i32 = 12;
        pub const MaxPageSource: i32 = 13;
        pub const CustomSource: i32 = 14;
        pub const LastPaperSource: i32 = 14;
        pub const Upper: i32 = 0;
    };

    pub const PrinterState = enum {
        pub const Idle: i32 = 0;
        pub const Active: i32 = 1;
        pub const Aborted: i32 = 2;
        pub const Error: i32 = 3;
    };

    pub const OutputFormat = enum {
        pub const NativeFormat: i32 = 0;
        pub const PdfFormat: i32 = 1;
    };

    pub const PrintRange = enum {
        pub const AllPages: i32 = 0;
        pub const Selection: i32 = 1;
        pub const PageRange: i32 = 2;
        pub const CurrentPage: i32 = 3;
    };

    pub const Unit = enum {
        pub const Millimeter: i32 = 0;
        pub const Point: i32 = 1;
        pub const Inch: i32 = 2;
        pub const Pica: i32 = 3;
        pub const Didot: i32 = 4;
        pub const Cicero: i32 = 5;
        pub const DevicePixel: i32 = 6;
    };

    pub const DuplexMode = enum {
        pub const DuplexNone: i32 = 0;
        pub const DuplexAuto: i32 = 1;
        pub const DuplexLongSide: i32 = 2;
        pub const DuplexShortSide: i32 = 3;
    };

};

File: binding/src/printsupport/libqprinter.cpp
#include <QPageSize>
#include <QPagedPaintDevice>
#include <QPaintEngine>
#include <QPrintEngine>
#include <QPrinter>
#include <QPrinterInfo>
#include <QRectF>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <qprinter.h>
#include "libqprinter.h"
#include "libqprinter.hxx"

QPrinter* QPrinter_new() {
	 return new VirtualQPrinter();
}

QPrinter* QPrinter_new2(const QPrinterInfo* printer) {
	 return new VirtualQPrinter(*printer);
}

QPrinter* QPrinter_new3(int mode) {
	 return new VirtualQPrinter(static_cast<QPrinter::PrinterMode>(mode));
}

QPrinter* QPrinter_new4(const QPrinterInfo* printer, int mode) {
	 return new VirtualQPrinter(*printer, static_cast<QPrinter::PrinterMode>(mode));
}

int QPrinter_DevType(const QPrinter* self) {
	return self->devType();
}

void QPrinter_SetOutputFormat(QPrinter* self, int format) {
	self->setOutputFormat(static_cast<QPrinter::OutputFormat>(format));
}

int QPrinter_OutputFormat(const QPrinter* self) {
	return self->outputFormat();
}

void QPrinter_SetPdfVersion(QPrinter* self, int version) {
	self->setPdfVersion(static_cast<QPagedPaintDevice::PdfVersion>(version));
}

int QPrinter_PdfVersion(const QPrinter* self) {
	return self->pdfVersion();
}

void QPrinter_SetPrinterName(QPrinter* self, const libqt_string printerName) {
	self->setPrinterName(QString::fromUtf8(printerName.data, printerName.len));
}

libqt_string QPrinter_PrinterName(const QPrinter* self) {
	QString _ret = self->printerName();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

bool QPrinter_IsValid(const QPrinter* self) {
	return self->isValid();
}

void QPrinter_SetOutputFileName(QPrinter* self, const libqt_string outputFileName) {
	self->setOutputFileName(QString::fromUtf8(outputFileName.data, outputFileName.len));
}

libqt_string QPrinter_OutputFileName(const QPrinter* self) {
	QString _ret = self->outputFileName();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QPrinter_SetPrintProgram(QPrinter* self, const libqt_string printProgram) {
	self->setPrintProgram(QString::fromUtf8(printProgram.data, printProgram.len));
}

libqt_string QPrinter_PrintProgram(const QPrinter* self) {
	QString _ret = self->printProgram();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QPrinter_SetDocName(QPrinter* self, const libqt_string docName) {
	self->setDocName(QString::fromUtf8(docName.data, docName.len));
}

libqt_string QPrinter_DocName(const QPrinter* self) {
	QString _ret = self->docName();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QPrinter_SetCreator(QPrinter* self, const libqt_string creator) {
	self->setCreator(QString::fromUtf8(creator.data, creator.len));
}

libqt_string QPrinter_Creator(const QPrinter* self) {
	QString _ret = self->creator();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QPrinter_SetPageOrder(QPrinter* self, int pageOrder) {
	self->setPageOrder(static_cast<QPrinter::PageOrder>(pageOrder));
}

int QPrinter_PageOrder(const QPrinter* self) {
	return self->pageOrder();
}

void QPrinter_SetResolution(QPrinter* self, int resolution) {
	self->setResolution(resolution);
}

int QPrinter_Resolution(const QPrinter* self) {
	return self->resolution();
}

void QPrinter_SetColorMode(QPrinter* self, int colorMode) {
	self->setColorMode(static_cast<QPrinter::ColorMode>(colorMode));
}

int QPrinter_ColorMode(const QPrinter* self) {
	return self->colorMode();
}

void QPrinter_SetCollateCopies(QPrinter* self, bool collate) {
	self->setCollateCopies(collate);
}

bool QPrinter_CollateCopies(const QPrinter* self) {
	return self->collateCopies();
}

void QPrinter_SetFullPage(QPrinter* self, bool fullPage) {
	self->setFullPage(fullPage);
}

bool QPrinter_FullPage(const QPrinter* self) {
	return self->fullPage();
}

void QPrinter_SetCopyCount(QPrinter* self, int copyCount) {
	self->setCopyCount(copyCount);
}

int QPrinter_CopyCount(const QPrinter* self) {
	return self->copyCount();
}

bool QPrinter_SupportsMultipleCopies(const QPrinter* self) {
	return self->supportsMultipleCopies();
}

void QPrinter_SetPaperSource(QPrinter* self, int paperSource) {
	self->setPaperSource(static_cast<QPrinter::PaperSource>(paperSource));
}

int QPrinter_PaperSource(const QPrinter* self) {
	return self->paperSource();
}

void QPrinter_SetDuplex(QPrinter* self, int duplex) {
	self->setDuplex(static_cast<QPrinter::DuplexMode>(duplex));
}

int QPrinter_Duplex(const QPrinter* self) {
	return self->duplex();
}

libqt_list QPrinter_SupportedResolutions(const QPrinter* self) {
	return self->supportedResolutions();
}

libqt_list QPrinter_SupportedPaperSources(const QPrinter* self) {
	return self->supportedPaperSources();
}

void QPrinter_SetFontEmbeddingEnabled(QPrinter* self, bool enable) {
	self->setFontEmbeddingEnabled(enable);
}

bool QPrinter_FontEmbeddingEnabled(const QPrinter* self) {
	return self->fontEmbeddingEnabled();
}

QRectF* QPrinter_PaperRect(const QPrinter* self, int param1) {
	return new QRectF(self->paperRect(static_cast<QPrinter::Unit>(param1)));
}

QRectF* QPrinter_PageRect(const QPrinter* self, int param1) {
	return new QRectF(self->pageRect(static_cast<QPrinter::Unit>(param1)));
}

libqt_string QPrinter_PrinterSelectionOption(const QPrinter* self) {
	QString _ret = self->printerSelectionOption();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QPrinter_SetPrinterSelectionOption(QPrinter* self, const libqt_string printerSelectionOption) {
	self->setPrinterSelectionOption(QString::fromUtf8(printerSelectionOption.data, printerSelectionOption.len));
}

bool QPrinter_NewPage(QPrinter* self) {
	return self->newPage();
}

bool QPrinter_Abort(QPrinter* self) {
	return self->abort();
}

int QPrinter_PrinterState(const QPrinter* self) {
	return self->printerState();
}

QPaintEngine* QPrinter_PaintEngine(const QPrinter* self) {
	return self->paintEngine();
}

QPrintEngine* QPrinter_PrintEngine(const QPrinter* self) {
	return self->printEngine();
}

void QPrinter_SetFromTo(QPrinter* self, int fromPage, int toPage) {
	self->setFromTo(fromPage, toPage);
}

int QPrinter_FromPage(const QPrinter* self) {
	return self->fromPage();
}

int QPrinter_ToPage(const QPrinter* self) {
	return self->toPage();
}

void QPrinter_SetPrintRange(QPrinter* self, int range) {
	self->setPrintRange(static_cast<QPrinter::PrintRange>(range));
}

int QPrinter_PrintRange(const QPrinter* self) {
	return self->printRange();
}

// Base class handler implementation
int QPrinter_QBaseDevType(const QPrinter* self) {
	auto* vqprinter = dynamic_cast<const VirtualQPrinter*>(self);
	if (vqprinter && vqprinter->isVirtualQPrinter) {
vqprinter->setQPrinter_DevType_IsBase(true);
	return vqprinter->devType();
}
}

// Auxiliary method to allow providing re-implementation
void QPrinter_OnDevType(const QPrinter* self, intptr_t slot) {
	auto* vqprinter = dynamic_cast<const VirtualQPrinter*>(self);
	if (vqprinter && vqprinter->isVirtualQPrinter) {
vqprinter->setQPrinter_DevType_Callback(reinterpret_cast<VirtualQPrinter::QPrinter_DevType_Callback>(slot));
}
}

// Base class handler implementation
bool QPrinter_QBaseNewPage(QPrinter* self) {
	auto* vqprinter = dynamic_cast<VirtualQPrinter*>(self);
	if (vqprinter && vqprinter->isVirtualQPrinter) {
vqprinter->setQPrinter_NewPage_IsBase(true);
	return vqprinter->newPage();
}
}

// Auxiliary method to allow providing re-implementation
void QPrinter_OnNewPage(QPrinter* self, intptr_t slot) {
	auto* vqprinter = dynamic_cast<VirtualQPrinter*>(self);
	if (vqprinter && vqprinter->isVirtualQPrinter) {
vqprinter->setQPrinter_NewPage_Callback(reinterpret_cast<VirtualQPrinter::QPrinter_NewPage_Callback>(slot));
}
}

// Base class handler implementation
QPaintEngine* QPrinter_QBasePaintEngine(const QPrinter* self) {
	auto* vqprinter = dynamic_cast<const VirtualQPrinter*>(self);
	if (vqprinter && vqprinter->isVirtualQPrinter) {
vqprinter->setQPrinter_PaintEngine_IsBase(true);
	return vqprinter->paintEngine();
}
}

// Auxiliary method to allow providing re-implementation
void QPrinter_OnPaintEngine(const QPrinter* self, intptr_t slot) {
	auto* vqprinter = dynamic_cast<const VirtualQPrinter*>(self);
	if (vqprinter && vqprinter->isVirtualQPrinter) {
vqprinter->setQPrinter_PaintEngine_Callback(reinterpret_cast<VirtualQPrinter::QPrinter_PaintEngine_Callback>(slot));
}
}

// Derived class handler implementation
int QPrinter_Metric(const QPrinter* self, int param1) {
	auto* vqprinter = dynamic_cast<const VirtualQPrinter*>(self);
	if (vqprinter && vqprinter->isVirtualQPrinter) {
	return vqprinter->metric(static_cast<QPaintDevice::PaintDeviceMetric>(param1));
	} else {
	return self->QPrinter::metric(static_cast<QPaintDevice::PaintDeviceMetric>(param1));
}
}

// Base class handler implementation
int QPrinter_QBaseMetric(const QPrinter* self, int param1) {
	auto* vqprinter = dynamic_cast<const VirtualQPrinter*>(self);
	if (vqprinter && vqprinter->isVirtualQPrinter) {
vqprinter->setQPrinter_Metric_IsBase(true);
	return vqprinter->metric(static_cast<QPaintDevice::PaintDeviceMetric>(param1));
}
}

// Auxiliary method to allow providing re-implementation
void QPrinter_OnMetric(const QPrinter* self, intptr_t slot) {
	auto* vqprinter = dynamic_cast<const VirtualQPrinter*>(self);
	if (vqprinter && vqprinter->isVirtualQPrinter) {
vqprinter->setQPrinter_Metric_Callback(reinterpret_cast<VirtualQPrinter::QPrinter_Metric_Callback>(slot));
}
}

// Derived class handler implementation
void QPrinter_SetEngines(QPrinter* self, QPrintEngine* printEngine, QPaintEngine* paintEngine) {
	auto* vqprinter = dynamic_cast<VirtualQPrinter*>(self);
	if (vqprinter && vqprinter->isVirtualQPrinter) {
	vqprinter->setEngines(printEngine, paintEngine);
	} else {
	self->QPrinter::setEngines(printEngine, paintEngine);
}
}

// Base class handler implementation
void QPrinter_QBaseSetEngines(QPrinter* self, QPrintEngine* printEngine, QPaintEngine* paintEngine) {
	auto* vqprinter = dynamic_cast<VirtualQPrinter*>(self);
	if (vqprinter && vqprinter->isVirtualQPrinter) {
vqprinter->setQPrinter_SetEngines_IsBase(true);
	vqprinter->setEngines(printEngine, paintEngine);
}
}

// Auxiliary method to allow providing re-implementation
void QPrinter_OnSetEngines(QPrinter* self, intptr_t slot) {
	auto* vqprinter = dynamic_cast<VirtualQPrinter*>(self);
	if (vqprinter && vqprinter->isVirtualQPrinter) {
vqprinter->setQPrinter_SetEngines_Callback(reinterpret_cast<VirtualQPrinter::QPrinter_SetEngines_Callback>(slot));
}
}

void QPrinter_Delete(QPrinter* self) {
    delete self;
}

